pkg/jobstore/boltdb/context.go

```diff
@@ -0,0 +1,66 @@
+package boltjobstore
+
+import (
+	"context"
+
+	"github.com/bacalhau-project/bacalhau/pkg/jobstore"
+	bolt "go.etcd.io/bbolt"
+)
+
+// contextKey is a custom type to avoid key collisions in context values.
+type contextKey int
+
+// txContextKey is the key used to store the transaction context in the context.
+const txContextKey contextKey = 0
+
+// txContext extends context.Context with transaction specific functionality.
+type txContext struct {
+	context.Context
+	tx         *bolt.Tx
+	cancelFunc context.CancelFunc
+}
+
+// newTxContext creates a new transactional context for a BoltDB transaction.
+// It embeds a standard context and manages transaction commit/rollback based on the context's lifecycle.
+func newTxContext(ctx context.Context, tx *bolt.Tx) *txContext {
+	innerCtx, cancelFunc := context.WithCancel(context.WithValue(ctx, txContextKey, tx))
+	txCtx := &txContext{
+		Context:    innerCtx,
+		tx:         tx,
+		cancelFunc: cancelFunc,
+	}
+	// Start a goroutine that listens for the context's Done channel.
+	go func() {
+		<-innerCtx.Done()
+		// If the transaction has not been committed when the context is done, rollback the transaction.
+		// In boltdb transaction, db is nil if the transaction is already committed or rolled back.
+		if txCtx.tx.DB() != nil {
+			_ = txCtx.tx.Rollback()
+		}
+	}()
```

Comment:
Few concerns:
1. Don't ignore the error if the rollback fails, at least log it.
2. When `txContext.Commit()/Rollback()` is called we commit/rollback the transaction, then cancel the inner context, if the DB is not nil we then rollback the transaction. Since boltdb's transaction implementation closes the connection on commit or rollback this is fine - But we've tightly coupled the logic here to the internals of boltdb, which is unfortunate and unavoidable given this approach. 
3. When the parent context is canceled we rollback the transaction, but depending on how this is used we may call `Commit` or `Rollback` again later which will result in `ErrTxClosed` errors. Again unfortunate but unavoidable.

Lets address 1 with a code change, and make 2-3 explicitly clear in comments

Reply:
improved things here https://github.com/bacalhau-project/bacalhau/pull/3996/commits/2aa9913b48bd13efd934d5af2fe311119b45f5ce, but I've also realized boltdb transactions are not thread safe and fixed it here https://github.com/bacalhau-project/bacalhau/pull/3996/commits/99eec412c1fc12a7481514ebc488a6413f602981